### ISSUE TITLE
periodicTask take up a lot of CPU time.

### ISSUE THREAD

**Issue by hiyanxu** [pyroscope-io/pyroscope#294]
![image](https://user-images.githubusercontent.com/15027927/126619675-ac895823-7c88-4a5b-9cc3-4b2d762a7d56.png)
The pyroscope will take up a lot of CPU time when I start up the pyroscope-server. 
By CPU ppof I found that periodicTask take up a lot of CPU time.
How can it be optimized?

**Comment by kolesnikovae** [comment]
Thank you for reporting the issue! 

Badgerdb GC call may be quite expensive. Recently, we made some improvements that should alleviate this problem, but those have not been released yet.

Could you please clarify the following:
 1. Pyroscope version.
 2. DB size on disk (e.g. `du -sh`).
 3. What is the pyroscope server CPU usage from OS perspective (%)?
 4. Does the problem persist during the server run or you observe this only at start?

**Comment by hiyanxu** [comment]
We integrated the pyroscope-server  into a platform called bianque.
So the pid of bianque is:
![image](https://user-images.githubusercontent.com/15027927/126629799-ead8ac62-af83-4fdc-a14c-178c7f0d09d0.png)


Pyroscope version: lastest branch main

DB size on disk: 
![image](https://user-images.githubusercontent.com/15027927/126629278-a840ff65-a60c-4f48-b734-b86fa45d9bbd.png)

CPU usage：
The machine is 8 vCPU 32 GiB.
![image](https://user-images.githubusercontent.com/15027927/126630353-b61c5db7-741b-4a30-8471-793c491c2817.png)
Sometimes CPU usage will be 150%.

Not only started, the problem persist during the server run.

Is it my configuration error?

**Comment by hiyanxu** [comment]
Is this a bug? Why call RunValueLogGC in infinite loop?
![image](https://user-images.githubusercontent.com/15027927/126633004-a8097dfb-6c95-4174-862a-7293da52b0d4.png)

When i delete the infinite loop(for{}), the CPU pprof is:
![image](https://user-images.githubusercontent.com/15027927/126635498-4b2e477b-1ba5-4b2e-9a8a-d26048f2af26.png)

**Comment by kolesnikovae** [comment]
Thank you for the info, it's very useful.

Regarding the code snippet – this should not be a problem and the loop is not supposed to be an infinite one:
```
// If a call to RunValueLogGC results in no rewrites, then an ErrNoRewrite is
// thrown indicating that the call resulted in no file rewrites.
```
And from [the official documentation](https://dgraph.io/docs/badger/get-started/#garbage-collection):

> This method is designed to do garbage collection while Badger is online. Along with randomly picking a file, it uses statistics generated by the LSM-tree compactions to pick files that are likely to lead to maximum space reclamation. It is recommended to be called during periods of low activity in your system, or periodically. **One call would only result in removal of at max one log file. As an optimization, you could also immediately re-run it whenever it returns nil error (indicating a successful value log GC)**, as shown below.

Thus, once all value log files are GC'ed, the `badgerGCTask` call returns. Deleting the loop effectively disables GC (almost), which leads to database bloating. 

----

I believe, it is very unlikely that this is has something to do with configuration. Could you also please send:
- output of `pyroscope version`.
- pyroscope server configuration (args, env vars, config file).

It would be also very helpful to take a look on memory consumption (it is surprisingly low judging by the screenshots) and pyroscope metrics: especially related to caching.

**Comment by hiyanxu** [comment]
pyroscope version: v0.0.35
config:
![image](https://user-images.githubusercontent.com/15027927/126729534-e1d66e45-dacf-4c6e-b310-b21d617b7f7c.png)

Thanks again for your attention!

**Comment by kolesnikovae** [comment]
Thank you! What caught my eye is that cache eviction options are set to 0 – this may cause OOM kill. Defaults are:
```
CacheEvictThreshold: 0.25
CacheEvictVolume:    0.33 
```

I suspect that badger GC consumes so much CPU time because of too many badger key versions are created for cached items. Upcoming release (0.0.36) brings changes to LFU cache and persistence model, it should fix the issue. Before the release you can verify if the problem persists in [HEAD](https://github.com/pyroscope-io/pyroscope/commit/a9138ba71033d779730aab123e73a6ac64d4d51f) (may be not stable).

**Comment by hiyanxu** [comment]
Thank you. I'll try.

**Comment by hiyanxu** [comment]
Thanks for your answer, I tried it.
It was effective at the beginning, but the next day I found that it took up a lot of CPU, even more than 200%.
I observed CPU pprof. The func runtime. gcBgMarkWorker is max.
![image](https://user-images.githubusercontent.com/15027927/126893273-bc7c04e6-064e-4365-88b7-9e10a717c02b.png)
I looked further at the number of memory inuse_objects, ingestHandler produced objects up to 98%.
![image](https://user-images.githubusercontent.com/15027927/126893399-382b1f76-1133-4012-822e-2f91f2e12369.png)

So is there a problem?

**Comment by kolesnikovae** [comment]
Thank you @hiyanxu, I very appreciate your help!

> It was effective at the beginning, but the next day I found that it took up a lot of CPU, even more than 200%.
I observed CPU pprof. The func runtime. gcBgMarkWorker is max.

I'm wondering if such a high utilisation is observed periodically or is constant. In Go, GC takes up to 25% of CPU, so that would mean that GC works all the time.

> I looked further at the number of memory inuse_objects, ingestHandler produced objects up to 98%.

This is fine. Apart from data ingestion and queries, there should be no other causes for allocations.

----

Can we take a look at pyroscope server metrics?

**Comment by hiyanxu** [comment]
Ok, thank you.
The first question:
> I'm wondering if such a high utilisation is observed periodically or is constant. In Go, GC takes up to 25% of CPU, so that would mean that GC works all the time.

It grows gradually. 
More specific information, I'll provide it tomorrow.

**Comment by kolesnikovae** [comment]
The number of GCs is definitely abnormal. Is it possible that GC is invoked explicitly (`runtime.GC()`) in the code?
```
go_gc_duration_seconds_count 64026
```

Can I also take a look at `pyroscope.server.alloc_space` profile (from Pyroscope UI)? I was trying to reproduce the issue but couldn't. How many clients do you have?

**Comment by hiyanxu** [comment]
GC is not invoked explicitly in the code.
`pyroscope.server` profile was disabled.

How many clients: Just one.

**But more importantly, over time, the agent takes more and more CPU times.**
We have three servers. `basic-exp-logic1` and `basic-exp-logic2` was profiling by  pyroscope, it consumes more CPU.
![image](https://user-images.githubusercontent.com/15027927/127090432-aa2843e4-9f9f-4aeb-bd78-24310e21ca18.png)

Compare by Pyroscope UI. `ProfileSession taskSnapshots()` is taking more and more CPU.
![image](https://user-images.githubusercontent.com/15027927/127090642-d679c812-3bd1-4e68-921b-adeabb56a2a7.png)

Server: 4vcpu8G.
Config: 
  DisableGCRuns: true
  SampleRate: 20

So, is it normal?
Thank you!

**Comment by kolesnikovae** [comment]
Hi @hiyanxu, I'd like to let you know that I refactored some of problematic functions: https://github.com/pyroscope-io/pyroscope/pull/303 - the work is ongoing, thank you for your help! :)

Regarding `takeSnapshots`: its performance mostly determined by the runtime: the more go routines exist and the bigger the heap, the more CPU time `takeSnapshots` call takes. However, the PR above should improve its performance. Anyway, the overhead of profiling in the picture is dumbfounding! It is supposed to be less than 5%.

**Comment by hiyanxu** [comment]
Ok, thank you for your attention!
I'll try to find a solution again about `takeSnapshots`.

Thanks again.

**Comment by kolesnikovae** [comment]
I created a separate issue regarding Go GC CPU consumption https://github.com/pyroscope-io/pyroscope/issues/306. There is a way to temporary workaround the problem:
> A workaround for deployed servers that experience the problem is setting very aggressive cache eviction with `-cache-evict-threshold` option, e.g, `0.1` or less (depending on the available memory): the option specifies percent of consumed RAM at which cached items are getting evicted, the amount of evicted items is controlled by `-cache-evict-volume`.

The initial problem with Badger GC CPU consumption fixed in 0.0.36 (not released yet), this issue is to be closed after the release.

**Comment by kolesnikovae** [comment]
We've released v0.0.36 with the fix and some optimizations that should resolve https://github.com/pyroscope-io/pyroscope/issues/307. 

@hiyanxu, could you please confirm that the initial problem is solved.

**Comment by hiyanxu** [comment]
I'm sorry to tell you that I can't continue the original project because of some changes in my work.
I have communicated with my original colleagues. If they continue the project, I will inform you whether it is effective.
Sorry~
